week-1-ts-ecosystem/w01e02-enter-the-world-of-typescript/src/framework/dom-creators.ts

```diff
@@ -0,0 +1,16 @@
+import { TagName } from "types";
+
+function createElement(tagName: TagName = "div", className: string = ""): HTMLElement | HTMLParagraphElement | HTMLDivElement {
+  const element = document.createElement(tagName);
+
+  if (className.length) {
+    element.className = className;
+  }
+
+  return element;
+}
+
+export const p = (className: string = ""): HTMLElement => createElement("p", className);
+export const div = (className: string = ""): HTMLElement => createElement("div", className);
```

Comment:
a tutaj z kolei to zwraca `HTMLDivElement`

Reply:
Tak jak masz teraz @teziovsky zapisane to się zgadza.
Tak się nie da, ale da się inaczej:

![image](https://user-images.githubusercontent.com/25398949/192050206-799dbb9f-3dd3-4abb-b9ce-244d480fdc4b.png)


1. Wyciągnij sobie wszystkie te `TagName`, tutaj ponieważ nigdzie indziej ich nie potrzebujesz, oraz jak zachowasz je tak jak teraz tutaj, to od razu widzisz co możesz sobie użyć w `createElement`

2. Jeśli w funkcji `createElement` użyjesz typu generycznego który rozszerza Twój przygotowany `TagName` to zadzieje się magic, bo `TypeScript` będzie w stanie określić co dokładnie leci do `document.createElement` i nawet jakbyś nie postawił typu w linii `:15` że to `HTMLParagraphElement` - to `TS` i tak to wywnioskuje 🤯 !

3. Staraj się unikać takich "Trywialnych" typowań, te parametry `className` mogą być tylko `string` bo taką mają wartość domyślną, jako pusty `string` i nie ma innej opcji i to widać na 1 rzut oka 😃

Reply:
Faktycznie, nie pomyślałem o typie generycznym. Poprawiłem w [d7f2c53 – tym commicie](https://github.com/michaljabi/js-na-frontach-zadania/pull/3/commits/d7f2c53fb9e02700ca73ef265aea273c45aa03da).